ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeContributionPool.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.subscribers.Subscribers;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.datastructures.blocks.blockbody.versions.altair.SyncAggregate;
+import tech.pegasys.teku.spec.datastructures.operations.versions.altair.SignedContributionAndProof;
+import tech.pegasys.teku.spec.datastructures.operations.versions.altair.SyncCommitteeContribution;
+import tech.pegasys.teku.statetransition.OperationPool.OperationAddedSubscriber;
+import tech.pegasys.teku.statetransition.validation.InternalValidationResult;
+import tech.pegasys.teku.util.time.channels.SlotEventsChannel;
+
+public class SyncCommitteeContributionPool implements SlotEventsChannel {
+
+  private final Spec spec;
+  private final SignedContributionAndProofValidator validator;
+  private final Subscribers<OperationAddedSubscriber<SignedContributionAndProof>> subscribers =
+      Subscribers.create(true);
+
+  /** contribution.slot -> contribution.block -> contribution.subcommitteeIndex -> contribution */
+  private final NavigableMap<UInt64, Map<Bytes32, Map<Integer, SyncCommitteeContribution>>>
+      contributionsBySlotAndBlockRoot = new TreeMap<>();
+
+  public SyncCommitteeContributionPool(
+      final Spec spec, final SignedContributionAndProofValidator validator) {
+    this.spec = spec;
+    this.validator = validator;
+  }
+
+  public void subscribeOperationAdded(
+      OperationAddedSubscriber<SignedContributionAndProof> subscriber) {
+    subscribers.subscribe(subscriber);
+  }
+
+  public SafeFuture<InternalValidationResult> add(
+      final SignedContributionAndProof signedContributionAndProof) {
+    return validator
+        .validate(signedContributionAndProof)
+        .thenPeek(
+            result -> {
+              if (result.isAccept()) {
+                doAdd(signedContributionAndProof.getMessage().getContribution());
+                subscribers.forEach(
+                    subscriber -> subscriber.onOperationAdded(signedContributionAndProof, result));
+              }
+            });
+  }
+
+  private synchronized void doAdd(final SyncCommitteeContribution contribution) {
+    final int subcommitteeIndex = contribution.getSubcommitteeIndex().intValue();
+    contributionsBySlotAndBlockRoot
+        .computeIfAbsent(contribution.getSlot(), __ -> new HashMap<>())
+        .computeIfAbsent(contribution.getBeaconBlockRoot(), __ -> new HashMap<>())
+        .compute(
+            subcommitteeIndex,
+            (subcommittee, existingContribution) ->
+                betterContribution(existingContribution, contribution));
+  }
+
+  private SyncCommitteeContribution betterContribution(
+      final SyncCommitteeContribution a, SyncCommitteeContribution b) {
+    if (a == null) {
+      return b;
+    } else if (b == null) {
+      return a;
+    } else {
```

Comment:
biggest of nitty nits: don't really need the else, the final return could be top level

Reply:
Yeah but technically you could skip all the `else` in here.  It's personal preference but I tend to like making it clear the three cases are linked with things like this.